src/graphics/aurora/flashfont.cpp

```diff
@@ -0,0 +1,208 @@
+}
+
+FlashFont::~FlashFont() {
+	cairo_font_face_destroy(_font);
+}
+
+cairo_status_t FlashFont::renderGlyph(cairo_scaled_font_t *scaled_font, unsigned long glyph, cairo_t *cr,
+									  cairo_text_extents_t *extents) {
+	FlashFont *font = reinterpret_cast<FlashFont *>(
+			cairo_font_face_get_user_data(
+					cairo_scaled_font_get_font_face(scaled_font),
+					&kUserDataKey
+			)
+	);
+
+	info("%i", static_cast<uint16>(glyph));
+
+	::Aurora::GFXCharacter::Glyph flashGlyph = font->_glyphs[glyph].glyph;
+
+	const double ascent = font->_ascent / 20.0;
+	const double descent = font->_descent / 20.0;
+
+	const double ascentRatio = ascent / (ascent + descent);
+
+	cairo_new_path(cr);
+
+	double x = 0, y = 0;
+	for (const auto &shapeRecord : flashGlyph.shapeRecords) {
+		if (shapeRecord.move.deltaX || shapeRecord.move.deltaY) {
+			x = static_cast<double>(shapeRecord.move.deltaX) / 1024.0;
+			y = static_cast<double>(shapeRecord.move.deltaY) / (ascent + descent);
+			cairo_move_to(cr, x, y + ascentRatio);
+		} else if (shapeRecord.straightEdge.deltaX || shapeRecord.straightEdge.deltaY) {
+			x += static_cast<double>(shapeRecord.straightEdge.deltaX) / 1024.0;
+			y += static_cast<double>(shapeRecord.straightEdge.deltaY) / (ascent + descent);
+			cairo_line_to(cr, x, y + ascentRatio);
+		} else if (shapeRecord.curvedEdge.anchorDeltaX || shapeRecord.curvedEdge.anchorDeltaY) {
+			double x0, y0;
+			cairo_get_current_point(cr, &x0, &y0);
+
+			double x1 = x + static_cast<double>(shapeRecord.curvedEdge.controlDeltaX) / 1024.0;
+			double y1 = y + static_cast<double>(shapeRecord.curvedEdge.controlDeltaY) / (ascent + descent);
+			double x2 = x + (static_cast<double>(shapeRecord.curvedEdge.anchorDeltaX)
+			        + static_cast<double>(shapeRecord.curvedEdge.controlDeltaX)) / 1024.0;
+			double y2 = y + (static_cast<double>(shapeRecord.curvedEdge.anchorDeltaY)
+			        + static_cast<double>(shapeRecord.curvedEdge.controlDeltaY)) / (ascent + descent);
+
+			x = x2;
+			y = y2;
+
+			y1 += ascentRatio;
+			y2 += ascentRatio;
+
+			cairo_curve_to(
+					cr,
+					(x0 + 2.0 * x1) / 3.0,
+					(y0 + 2.0 * y1) / 3.0,
+					(x2 + 2.0 * x1) / 3.0,
+					(y2 + 2.0 * y1) / 3.0,
+					x2,
+					y2
+			);
+		}
+	}
+
+	cairo_close_path(cr);
+	cairo_fill(cr);
+
+	return CAIRO_STATUS_SUCCESS;
+}
+
+cairo_status_t FlashFont::textToGlyphs(cairo_scaled_font_t *scaled_font, const char *utf8, int utf8_len,
+	cairo_glyph_t **glyphs, int *num_glyphs, cairo_text_cluster_t **UNUSED(clusters), int *UNUSED(num_clusters),
+	cairo_text_cluster_flags_t *UNUSED(cluster_flags)) {
+	FlashFont *font = reinterpret_cast<FlashFont *>(
+			cairo_font_face_get_user_data(
+					cairo_scaled_font_get_font_face(scaled_font),
+					&kUserDataKey
+			)
+	);
+
+	const std::map<uint16, Glyph> &glyphmap = font->_glyphs;
+	const std::vector<::Aurora::GFXCharacter::KerningCode> &kernings = font->_kernings;
+
+	(*glyphs) = cairo_glyph_allocate(utf8_len);
+	(*num_glyphs) = utf8_len;
```

Comment:
You don't need these parentheses here.

But in fact, since we're C++, you might just make the method take a reference instead of a pointer. Especially since then no one can give it a nullptr (which you don't check for, so that would explode).

Reply:
(Oh, right, since the function's signature is predetermined by Cairo, you of course can't make those references. My bad :P)